In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sea

In [3]:
train = pd.read_csv("D:/Kaggle/Housing_Prices/kaggle-housing-prices-/train.csv")
print(train.shape)
test = pd.read_csv("D:/Kaggle/Housing_Prices/kaggle-housing-prices-/test.csv")
SalePrice = train['SalePrice']
train = train.drop(columns = ['SalePrice'])
df = pd.concat([train, test])
index = df['Id']


(1460, 81)


In [4]:
print(train.shape)
print(test.shape)
print(df.shape)


(1460, 80)
(1459, 80)
(2919, 80)


In [35]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  2919.000000  2919.000000  2433.000000    2919.000000  2919.000000   
mean   1460.000000    57.137718    69.305795   10168.114080     6.089072   
std     842.787043    42.517628    23.344905    7886.996359     1.409947   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     730.500000    20.000000    59.000000    7478.000000     5.000000   
50%    1460.000000    50.000000    68.000000    9453.000000     6.000000   
75%    2189.500000    70.000000    80.000000   11570.000000     7.000000   
max    2919.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  2919.000000  2919.000000   2919.000000  2896.000000  2918.000000  ...   
mean      5.564577  1971.312778   1984.264474   102.201312   441.423235  ...   
std       1.113131    30.291442     20.894344   179.334253   455.610826  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.500000   1965.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1993.000000     0.000000   368.500000  ...   
75%       6.000000  2001.000000   2004.000000   164.000000   733.000000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  2918.000000  2919.000000  2919.000000    2919.000000  2919.000000   
mean    472.874572    93.709832    47.486811      23.098321     2.602261   
std     215.394815   126.526589    67.575493      64.244246    25.188169   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     320.000000     0.000000     0.000000       0.000000     0.000000   
50%     480.000000     0.000000    26.000000       0.000000     0.000000   
75%     576.000000   168.000000    70.000000       0.000000     0.000000   
max    1488.000000  1424.000000   742.000000    1012.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  2919.000000  2919.000000   2919.000000  2919.000000  2919.000000  
mean     16.062350     2.251799     50.825968     6.213087  2007.792737  
std      56.184365    35.663946    567.402211     2.714762     1.314964  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     4.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     576.000000   800.000000  17000.000000    12.000000  2010.000000  

[8 rows x 37 columns]

In [6]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

['MSZoning', 'LotFrontage', 'Alley', 'Utilities', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Electrical', 'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType']


In [7]:
df[Cols_Null].describe()

LotFrontage   MasVnrArea   BsmtFinSF1   BsmtFinSF2    BsmtUnfSF  \
count  2433.000000  2896.000000  2918.000000  2918.000000  2918.000000   
mean     69.305795   102.201312   441.423235    49.582248   560.772104   
std      23.344905   179.334253   455.610826   169.205611   439.543659   
min      21.000000     0.000000     0.000000     0.000000     0.000000   
25%      59.000000     0.000000     0.000000     0.000000   220.000000   
50%      68.000000     0.000000   368.500000     0.000000   467.000000   
75%      80.000000   164.000000   733.000000     0.000000   805.500000   
max     313.000000  1600.000000  5644.000000  1526.000000  2336.000000   

       TotalBsmtSF  BsmtFullBath  BsmtHalfBath  GarageYrBlt   GarageCars  \
count  2918.000000   2917.000000   2917.000000  2760.000000  2918.000000   
mean   1051.777587      0.429894      0.061364  1978.113406     1.766621   
std     440.766258      0.524736      0.245687    25.574285     0.761624   
min       0.000000      0.000000      0.000000  1895.000000     0.000000   
25%     793.000000      0.000000      0.000000  1960.000000     1.000000   
50%     989.500000      0.000000      0.000000  1979.000000     2.000000   
75%    1302.000000      1.000000      0.000000  2002.000000     2.000000   
max    6110.000000      3.000000      2.000000  2207.000000     5.000000   

        GarageArea  
count  2918.000000  
mean    472.874572  
std     215.394815  
min       0.000000  
25%     320.000000  
50%     480.000000  
75%     576.000000  
max    1488.000000

In [8]:
df[Cols_Null].isnull().sum()

MSZoning           4
LotFrontage      486
Alley           2721
Utilities          2
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
KitchenQual        1
Functional         2
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
GarageFinish     159
GarageCars         1
GarageArea         1
GarageQual       159
GarageCond       159
PoolQC          2909
Fence           2348
MiscFeature     2814
SaleType           1
dtype: int64

Reading the data description for a lot of variables NA is a category, ie NA -> No alley access, and not a 'missing value'. We will walk through each variable. 

First is MSZoning. NA is not defined in data scription, thus this is a missing value. 

In [29]:
print(df.MSZoning.unique())
print(df['MSZoning'].value_counts())

['RL' 'RM' 'C (all)' 'FV' 'RH']
RL         2269
RM          460
FV          139
RH           26
C (all)      25
Name: MSZoning, dtype: int64


In [30]:
df['MSZoning'].fillna(df['MSZoning'].mode()[0], inplace = True)
print(df.MSZoning.unique())

['RL' 'RM' 'C (all)' 'FV' 'RH']


Replace MSZoning with Dummy Variables

In [47]:
MSZone_dummy_features = pd.get_dummies(df['MSZoning'], drop_first=True)
df = pd.concat([df, MSZone_dummy_features], axis=1)
df.head()


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... SaleType SaleCondition FV  RH  RL  RM FV  RH  RL  \
0         Lvl    AllPub  ...       WD        Normal  0   0   1   0  0   0   1   
1         Lvl    AllPub  ...       WD        Normal  0   0   1   0  0   0   1   
2         Lvl    AllPub  ...       WD        Normal  0   0   1   0  0   0   1   
3         Lvl    AllPub  ...       WD       Abnorml  0   0   1   0  0   0   1   
4         Lvl    AllPub  ...       WD        Normal  0   0   1   0  0   0   1   

   RM  
0   0  
1   0  
2   0  
3   0  
4   0  

[5 rows x 88 columns]

In [49]:
df.shape
df.drop(['MSZoning'], inplace=True, axis=1)

remove Pool Area in addition to PoolQC, provides little value since out of 1460 obs only 7 have a pool. 

In [ ]:
df[Cols_Null].dtypes

### Update NA in LotFrontage with the mean of LotFrontages with similar LotAreas. 

In [ ]:
df['LotFrontage'].describe()

In [ ]:
df["LotFrontage"].fillna(df['LotFrontage'].mean(), inplace = True) 
df['LotFrontage'].mean()

In [ ]:
df.drop(['Alley'], axis=1, inplace = True)
df.drop(['PoolQC'], axis=1, inplace = True)
df.drop(['PoolArea'], axis=1, inplace = True)
df.drop(['Fence'], axis=1, inplace = True)
df.drop(['MiscFeature'], axis=1, inplace = True)

In [ ]:
df.isnull().sum()


In [ ]:
df.info()

In [ ]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

In [ ]:
df['MasVnrType'].describe

In [ ]:
df['MasVnrType'].fillna('None', inplace = True) 
df['MasVnrArea'].fillna('None', inplace = True) 
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

In [ ]:
df['GarageType'].fillna('No Garage', inplace = True) 
df['GarageYrBlt'].fillna('No Garage', inplace = True) 
df['GarageFinish'].fillna('No Garage', inplace = True) 
df['GarageQual'].fillna('No Garage', inplace = True) 
df['GarageCond'].fillna('No Garage', inplace = True) 
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

In [ ]:
df['FireplaceQu'].fillna('No Fireplace', inplace = True)

In [ ]:
df['Electrical'].fillna(df['Electrical'].mode()[0], inplace = True)
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)




Replace BSMT Bathrooms with No Basement if Basement Qual is NA

In [ ]:
df.loc[df['BsmtQual'].isnull() & df['BsmtFullBath'].eq(0), 'BsmtFullBath'] = 'No Basement'
df.loc[df['BsmtQual'].isnull() & df['BsmtHalfBath'].eq(0), 'BsmtHalfBath'] = 'No Basement'
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)


In [ ]:
df[Cols_Null].isnull().sum()

In [ ]:
df['BsmtFullBath'] = df['BsmtFullBath'].apply(str)
print(np.unique(df[['BsmtFullBath']]))

In [ ]:
df['BsmtHalfBath'] = df['BsmtHalfBath'].apply(str)
print(np.unique(df[['BsmtHalfBath']]))

In [ ]:
df['BsmtQual'].fillna('No Basement', inplace = True) 
df['BsmtCond'].fillna('No Basement', inplace = True) 
df['BsmtExposure'].fillna('No Basement', inplace = True) 
df['BsmtFinType1'].fillna('No Basement', inplace = True) 
df['BsmtFinType2'].fillna('No Basement', inplace = True) 

In [ ]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)
df.info()


Update categorical columns as strings and creat dummy variables